File: ambeth/Ambeth.IoC.Test/ioc/extendable/ExtendableBeanTest.cs
using De.Osthus.Ambeth.Testutil;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Factory;
using De.Osthus.Ambeth.Ioc.Extendable;
using De.Osthus.Ambeth.Log;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections;

namespace De.Osthus.Ambeth.Ioc.Extendable
{
    [TestClass]
    [TestModule(typeof(ExtendableBeanTestModule))]
    public class ExtendableBeanTest : AbstractIocTest
    {
        public class ExtendableBeanTestModule : IInitializingModule
        {
            public void AfterPropertiesSet(IBeanContextFactory beanContextFactory)
            {
                beanContextFactory.registerBean<ExtendableBean>("testExtensionPoint")
                        .propertyValue(ExtendableBean.P_EXTENDABLE_TYPE, typeof(ITestListenerExtendable))
                        .propertyValue(ExtendableBean.P_PROVIDER_TYPE, typeof(ITestListenerRegistry));

                beanContextFactory.registerBean<ExtendableBean>("testExtensionPoint2")
                        .propertyValue(ExtendableBean.P_EXTENDABLE_TYPE, typeof(ITestListenerExtendable2))
                        .propertyValue(ExtendableBean.P_PROVIDER_TYPE, typeof(ITestListenerRegistry2));
            }
        }

        public class DummyListener : ITestListener
        {
        }

        [LogInstance]
        public ILogger Log { private get; set; }

        [TestMethod]
        public void TestExtensionPoint()
        {
            InitManually(GetType());
            ITestListenerExtendable testListenerExtendable = BeanContext.GetService<ITestListenerExtendable>("testExtensionPoint");
            ITestListenerRegistry testListenerRegistry = BeanContext.GetService<ITestListenerRegistry>("testExtensionPoint");

            testListenerExtendable.AddTestListener(new DummyListener());

            ITestListener[] testListeners = testListenerRegistry.GetTestListeners();
            Assert.IsNotNull(testListeners);
            Assert.AreEqual(1, testListeners.Length);
        }

        [TestMethod]
        public void TestExtensionPointByType()
        {
            InitManually(GetType());
            ITestListenerExtendable2 testListenerExtendable = BeanContext.GetService<ITestListenerExtendable2>("testExtensionPoint2");
            ITestListenerRegistry2 testListenerRegistry = BeanContext.GetService<ITestListenerRegistry2>("testExtensionPoint2");

            testListenerExtendable.RegisterTestListener(new DummyListener(), typeof(IList));

            ITestListener testListener = testListenerRegistry.GetTestListener(typeof(ArrayList));
            Assert.IsNotNull(testListener);
        }
    }
}

File: jambeth/jAmbeth.IoC/test/de/osthus/ambeth/ioc/extendable/ExtendableBeanTest.java
package de.osthus.ambeth.ioc.extendable;

import java.util.List;

import junit.framework.Assert;

import org.junit.Test;

import de.osthus.ambeth.collections.ArrayList;
import de.osthus.ambeth.ioc.IInitializingModule;
import de.osthus.ambeth.ioc.extendable.ExtendableBeanTest.ExtendableBeanTestModule;
import de.osthus.ambeth.ioc.factory.IBeanContextFactory;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.testutil.AbstractIocTest;
import de.osthus.ambeth.testutil.TestModule;

@TestModule(ExtendableBeanTestModule.class)
public class ExtendableBeanTest extends AbstractIocTest
{
	public static class ExtendableBeanTestModule implements IInitializingModule
	{
		@Override
		public void afterPropertiesSet(IBeanContextFactory beanContextFactory) throws Throwable
		{
			beanContextFactory.registerBean("testExtensionPoint", ExtendableBean.class)
					.propertyValue(ExtendableBean.P_EXTENDABLE_TYPE, ITestListenerExtendable.class)
					.propertyValue(ExtendableBean.P_PROVIDER_TYPE, ITestListenerRegistry.class);

			beanContextFactory.registerBean("testExtensionPoint2", ExtendableBean.class)
					.propertyValue(ExtendableBean.P_EXTENDABLE_TYPE, ITestListenerExtendable2.class)
					.propertyValue(ExtendableBean.P_PROVIDER_TYPE, ITestListenerRegistry2.class);
		}
	}

	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Test
	public void testToString()
	{
		ITestListenerExtendable testListenerExtendable = beanContext.getService("testExtensionPoint", ITestListenerExtendable.class);
		ITestListenerRegistry testListenerRegistry = beanContext.getService("testExtensionPoint", ITestListenerRegistry.class);

		Assert.assertNotNull(testListenerExtendable.toString());
		Assert.assertNotNull(testListenerRegistry.toString());
	}

	@Test
	public void testExtensionPoint()
	{
		ITestListenerExtendable testListenerExtendable = beanContext.getService("testExtensionPoint", ITestListenerExtendable.class);
		ITestListenerRegistry testListenerRegistry = beanContext.getService("testExtensionPoint", ITestListenerRegistry.class);

		testListenerExtendable.addTestListener(new ITestListener()
		{
		});

		ITestListener[] testListeners = testListenerRegistry.getTestListeners();
		Assert.assertNotNull(testListeners);
		Assert.assertEquals(1, testListeners.length);
	}

	@Test
	public void testExtensionPointByType()
	{
		ITestListenerExtendable2 testListenerExtendable = beanContext.getService("testExtensionPoint2", ITestListenerExtendable2.class);
		ITestListenerRegistry2 testListenerRegistry = beanContext.getService("testExtensionPoint2", ITestListenerRegistry2.class);

		testListenerExtendable.registerTestListener(new ITestListener()
		{
		}, List.class);

		ITestListener testListener = testListenerRegistry.getTestListener(ArrayList.class);
		Assert.assertNotNull(testListener);
	}
}
